### PR TITLE
[Feature] Support EVA-MAE style

### PR DESCRIPTION
**Pull request by wangbo-zhao**
Support train vit-Base in EVA-MAE style.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by wangbo-zhao** [comment]
Working in progress. I am pretraing the model.

File: tests/test_models/test_algorithms/test_eva.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    decoder_num_heads=16,
+    predict_feature_dim=512,
+    mlp_ratio=4.,
+)
+loss = dict(type='CosineSimilarityLoss', shift_factor=1.0, scale_factor=1.0)
+head = dict(type='EVAPretrainHead', loss=loss)
+
+
+@pytest.mark.skipif(platform.system() == 'Windows', reason='Windows mem limit')
+def test_milan():
```
Comment (YuanLiuuuuuu) [inline tests/test_models/test_algorithms/test_eva.py]:
```suggestion
def test_eva():
```

File: mmselfsup/models/heads/eva_head.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from mmengine.model import BaseModule
+
+from mmselfsup.registry import MODELS
+
+
+@MODELS.register_module()
+class EVAPretrainHead(BaseModule):
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/heads/eva_head.py]: Better reuse ``MILANPretrainHead`` by adding ``mask`` param to the ``forward`` func.

**Comment by wangbo-zhao** [comment]
Preparing links for ckpts

File: configs/selfsup/eva/metafile.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      Training Resources: 16x A100-80G GPUs
+      Architecture:
+        - ViT
+    Paper:
+        URL: https://arxiv.org/abs/2211.07636
+        Title: "EVA: Exploring the Limits of Masked Visual Representation Learning at Scale"
+    README: configs/selfsup/eva/README.md
+
+Models:
+  - Name: eva_mae-sytle_vit-base-p16_16xb256-coslr-400e_1k
```
Comment (fangyixiao18) [inline configs/selfsup/eva/metafile.yaml]:
```suggestion
  - Name: eva-mae-style_vit-base-p16_16xb256-coslr-400e_in1k
```
please also check your all other config names

File: mmselfsup/models/losses/cosine_similarity_loss.py [hunk 4]
```diff
@@ -28,8 +28,10 @@ def __init__(self,
         self.shift_factor = shift_factor
         self.scale_factor = scale_factor
 
-    def forward(self, pred: torch.Tensor,
-                target: torch.Tensor) -> torch.Tensor:
+    def forward(self,
+                pred: torch.Tensor,
+                target: torch.Tensor,
+                mask: torch.Tensor = None) -> torch.Tensor:
```
Comment (fangyixiao18) [inline mmselfsup/models/losses/cosine_similarity_loss.py]:
```suggestion
                mask: Optional[torch.Tensor] = None) -> torch.Tensor:
```

File: mmselfsup/models/necks/mae_neck.py [hunk 5]
```diff
@@ -55,6 +55,7 @@ def __init__(self,
                  decoder_num_heads: int = 16,
                  mlp_ratio: int = 4,
                  norm_cfg: dict = dict(type='LN', eps=1e-6),
+                 predict_feature_dim: float = None,
```
Comment (fangyixiao18) [inline mmselfsup/models/necks/mae_neck.py]:
```suggestion
                 predict_feature_dim: Optional[float] = None,
```

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: mmselfsup/models/algorithms/eva.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+        Args:
+            inputs (List[torch.Tensor]): The input images.
+            data_samples (List[SelfSupDataSample]): All elements required
+                during the forward function.
+
+        Returns:
+            Dict[str, torch.Tensor]: A dictionary of loss components.
+        """
+        # ids_restore: the same as that in original repo, which is used
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/eva.py]: Better move this comment to the line above line#36

File: configs/selfsup/eva/metafile.yaml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        Weights:
+      - Type: Image Classification
+        Metadata:
+          Epochs: 100
+          Batch Size: 16384
+        Results:
+          - Task: Linear-probing
+            Dataset: ImageNet-1k
+            Metrics:
+              Top 1 Accuracy: 69.01
```
Comment (YuanLiuuuuuu) [inline configs/selfsup/eva/metafile.yaml]: Keep one decimal

File: configs/selfsup/eva/metafile.yaml (hunk 3 above)
Comment (wangbo-zhao) [reply configs/selfsup/eva/metafile.yaml]: sorry I forget the metafile. Ohter config names have be modified.

**Review by YuanLiuuuuuu — APPROVED** [review]

**Comment by YuanLiuuuuuu** [comment]
Please fix the UT. Anything else looks good to me

File: configs/selfsup/eva/README.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  </thead>
+  <tbody>
+  <tr>
+      <td rowspan="1">EVA</td>
+	    <td>ViT-B/16</td>
+	    <td>400</td>
+      <td>4096</td>
+      <td>83.72</td>
+      <td>69.02</td>
+      <td><a href='https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/configs/selfsup/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k.py'>config</a> | <a href='https://download.openmmlab.com/mmselfsup/1.x/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k_20221129-180922e8.pth'>model</a> | <a href='https://download.openmmlab.com/mmselfsup/1.x/milan/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k/milan_vit-base-p16_16xb256-amp-coslr-400e_in1k_20221123_112721.json'>log</a></td>
```
Comment (fangyixiao18) [inline configs/selfsup/eva/README.md]: update the links

File: configs/selfsup/eva/README.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	</tr>
+  </thead>
+  <tbody>
+  <tr>
+      <td rowspan="1">EVA</td>
+	    <td>ViT-B/16</td>
+	    <td>400</td>
+      <td>4096</td>
+      <td>83.72</td>
+      <td>69.02</td>
```
Comment (fangyixiao18) [inline configs/selfsup/eva/README.md:45]: linear and ft results are in wrong position

File: configs/selfsup/eva/metafile.yaml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+      Training Resources: 16x A100-80G GPUs
+      Architecture:
+        - ViT
+    Paper:
+        URL: https://arxiv.org/abs/2211.07636
+        Title: "EVA: Exploring the Limits of Masked Visual Representation Learning at Scale"
+    README: configs/selfsup/eva/README.md
+
+Models:
+  - Name: eva_mae-sytle_vit-base-p16_16xb256-coslr-400e_in1k
```
Comment (fangyixiao18) [inline configs/selfsup/eva/metafile.yaml]:
```suggestion
  - Name: eva-mae-style_vit-base-p16_16xb256-coslr-400e_in1k
```

File: configs/selfsup/eva/classification/vit-base-p16_ft-8xb128-coslr-100e_in1k.py [hunk 11]
```diff
@@ -0,0 +1,12 @@
+_base_ = '../../../benchmarks/classification/imagenet/vit-base-p16_ft-8xb128-coslr-100e_in1k.py'  # noqa: E501
+
+# model settings
+model = dict(
+    head=dict(init_cfg=[dict(type='TruncNormal', layer='Linear', std=0.02)
+                        ]),  # MAE sets std to 2e-5
```
Comment (tonysy) [inline configs/selfsup/eva/classification/vit-base-p16_ft-8xb128-coslr-100e_in1k.py]: better to move the serveral brackets in line-6/7 to line-5

**Review by tonysy — APPROVED** [review]
LGTM

File: configs/selfsup/eva/README.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
+# MILAN
```
Comment (fangyixiao18) [inline configs/selfsup/eva/README.md]: EVA

**Review by fangyixiao18 — APPROVED** [review]
